src/Http/Http/src/Internal/DefaultHttpResponse.cs

```diff
@@ -139,11 +143,22 @@ public override void Redirect(string location, bool permanent)
             Headers[HeaderNames.Location] = location;
         }
 
+        public override Task StartAsync(bool flush, CancellationToken token)
+        {
+            if (HttpResponseStartFeature == null)
+            {
+                return HttpResponseFeature.Body.FlushAsync(token);
```

Comment:
This feature as is doesn't seem too useful when you realize HttpResponseStartFeature.StartAsync() can be replaced with HttpResponseFeature.Body.FlushAsync. Updating IResponseBodyPipeFeature at least prevents unintuitive runtime failures from accessing the response pipe too early.

Reply:
If we didn't have this feature, we would do two flushes per response rather than one.

> Updating IResponseBodyPipeFeature at least prevents unintuitive runtime failures from accessing the response pipe too early.

Let's say we go with your suggestion of doing GetPipeAsync on the HttpResponse. Isn't it also unintuitive that getting the pipe makes headers unmodifiable? I can see something like this being fairly common.
 
```c#
var pipeWriter = await response.GetPipeAsync();
response.ContentLength = 11; // Broken
// do pipe things
```

Introducing pipes will introduce a new style of request/response operations. If we are clear that using the PipeReader/Writer directly will obtain memory from the PipeWriter, it will make sense why people need to call start to commit headers.

Reply:
I'm not against renaming this to CommitHeaders (was a proposed name before). This call is a combination of both doing response start and committing headers. Maybe Committing headers implies the response has started? 

@Tratcher as you advocated for the StartAsync api, from an API standpoint, wouldn't calling StartAsync not imply that headers were committed? It implies that headers are unmodifiable, but not necessarily written to the pipe.

Reply:
Response.StartAsync aligns with OnStarting and HasStarted, and directly triggers both of those.